frame/migrations/src/lib.rs

```diff
@@ -0,0 +1,488 @@
+};
+use frame_system::Pallet as System;
+use sp_runtime::Saturating;
+use sp_std::{boxed::Box, vec::Vec};
+
+const LOG: &'static str = "runtime::migrations";
+
+/// Points to the next migration to execute.
+#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode, scale_info::TypeInfo, MaxEncodedLen)]
+pub enum MigrationCursor<Cursor, BlockNumber> {
+	/// Points to the currently active migration and its cursor.
+	Active(ActiveCursor<Cursor, BlockNumber>),
+	/// Migration got stuck and cannot proceed.
+	Stuck,
+}
+
+impl<Cursor, BlockNumber> MigrationCursor<Cursor, BlockNumber> {
+	/// Maybe return self as an `ActiveCursor`.
+	pub fn as_active(&self) -> Option<&ActiveCursor<Cursor, BlockNumber>> {
+		match self {
+			MigrationCursor::Active(active) => Some(active),
+			MigrationCursor::Stuck => None,
+		}
+	}
+}
+
+impl<Cursor, BlockNumber> From<ActiveCursor<Cursor, BlockNumber>> for MigrationCursor<Cursor, BlockNumber> {
+	fn from(active: ActiveCursor<Cursor, BlockNumber>) -> Self {
+		MigrationCursor::Active(active)
+	}
+}
+
+/// Points to the currently active migration and its inner cursor.
+#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode, scale_info::TypeInfo, MaxEncodedLen)]
+pub struct ActiveCursor<Cursor, BlockNumber> {
+	index: u32,
+	inner_cursor: Option<Cursor>,
+	started_at: BlockNumber,
+}
+
+impl<Cursor, BlockNumber> ActiveCursor<Cursor, BlockNumber> {
+	/// Advance the cursor to the next migration.
+	pub(crate) fn advance(&mut self, current_block: BlockNumber) {
+		self.index.saturating_inc();
+		self.inner_cursor = None;
+		self.started_at = current_block;
+	}
+}
+
+/// A collection of migrations that must be executed in order.
+pub type MigrationsOf<T> = Vec<
+	Box<
+		dyn SteppedMigration<
+			Cursor = <T as Config>::Cursor,
+			Identifier = <T as Config>::Identifier,
+		>,
+	>,
+>;
+
+/// Convenience alias for [`MigrationCursor`].
+pub type CursorOf<T> =
+	MigrationCursor<<T as Config>::Cursor, <T as frame_system::Config>::BlockNumber>;
+
+/// Convenience alias for [`ActiveCursor`].
+pub type ActiveCursorOf<T> =
+	ActiveCursor<<T as Config>::Cursor, <T as frame_system::Config>::BlockNumber>;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type of the runtime.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// All the multi-block migrations to run.
+		///
+		/// Should only be updated in a runtime-upgrade once all the old migrations have completed.
+		/// (Check that `Cursor` is `None`).
+		type Migrations: Get<MigrationsOf<Self>>;
+
+		/// The cursor type that is shared across all migrations.
+		type Cursor: FullCodec + MaxEncodedLen + TypeInfo + Parameter;
+
+		/// The identifier type that is shared across all migrations.
```

Comment:
I am a bit skeptical of why we need these two. 

The cursors are unique across migrations, and the only reasonable "shared cursor" is `Vec<u8>`? 

As for the identifier, it can probably be fixed to something like `&[u8; 16]`.

Reply:
Saying this because I was under the impression that we concluded that each migration needs to provide its own means of encoding and decoding its cursor, which means this pallet only sees them as `Vec<u8>`.

Moreover, in this case, you could technically store `#[pallet::storage] type Cursors = ..` as a `Map<Identifier, OpaqueCursor>`.